src/transport/SessionManager.cpp

```diff
@@ -1336,6 +1318,33 @@ Optional<SessionHandle> SessionManager::FindSecureSessionForNode(ScopedNodeId pe
     return mrpSession != nullptr ? MakeOptional<SessionHandle>(*mrpSession) : Optional<SessionHandle>::Missing();
 }
 
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+void SessionManager::MarkSecureSessionDefunctForConnection(Transport::ActiveTCPConnectionState * conn, CHIP_ERROR conErr)
+{
+    // Mark the corresponding secure sessions as defunct
+    mSecureSessions.ForEachSession([&](auto session) {
+        if (session->IsActiveSession() && session->GetTCPConnection() == conn)
+        {
+            SessionHandle handle(*session);
+            // Notify the SessionConnection delegate of the connection
+            // closure.
+            if (mConnDelegate != nullptr)
+            {
+                mConnDelegate->OnTCPConnectionClosed(handle, conErr);
+            }
+
+            // Dis-associate the connection from session by setting it to a
+            // nullptr.
+            session->SetTCPConnection(nullptr);
+            // Mark session as defunct
+            session->MarkAsDefunct();
```

Comment:
So here's a question: Why are we doing MarkAsDefunct(), not MarkForEviction()?  Is there a way this session can still be OK when the connection has been closed?  I wouldn't think so...

Reply:
I think it can be marked for eviction. The original intent was to, potentially, re-establish the connection and link a defunct session if one existed with said peer. But, currently, for the sake of simplicity, when a new connection is set up, a corresponding new session is also set up. This might be better as we may not want to reuse a session that has been stale for long. But I thought that a defunct session could also be evicted, no? If not then MarkForEviction is probably the right option.

Reply:
A defunct session can absolutely be evicted, but it might be prioritized for eviction below other sessions.  So if we know for sure our session is never going to be usable, MarkForEviction will be better in terms of making sure we don't leave this session around while evicting a potentially-usable session.